src/backend/input.rs

```diff
@@ -0,0 +1,277 @@
+        slot: Option<TouchSlot>,
+        /// Absolute x-coordinate of the touch position.
+        x: f64,
+        /// Absolute y-coordinate of the touch position.
+        y: f64
+    },
+    /// Movement of a touch on the device surface to a given position (x, y).
+    ///
+    /// If the device has multi-touch capabilities a slot is given.
+    Motion {
+        /// `TouchSlot`, if the device has multi-touch capabilities
+        slot: Option<TouchSlot>,
+        /// Absolute x-coordinate of the final touch position after the motion.
+        x: f64,
+        /// Absolute y-coordinate of the final touch position after the motion.
+        y: f64 },
+    /// Stop of an event chain.
+    ///
+    /// If the device has multi-touch capabilities a slot is given.
+    Up {
+        /// `TouchSlot`, if the device has multi-touch capabilities
+        slot: Option<TouchSlot>
+    },
+    /// Cancel of an event chain. All previous events in the chain should be ignored.
+    ///
+    /// If the device has multi-touch capabilities a slot is given.
+    Cancel {
+        /// `TouchSlot`, if the device has multi-touch capabilities
+        slot: Option<TouchSlot>
+    },
+    /// Signals the end of a set of touchpoints at one device sample time.
+    Frame
+}
+
+/// Trait that describes objects providing a source of input events. All input backends
+/// need to implemenent this and provide the same base gurantees about the presicion of
+/// given events.
+pub trait InputBackend {
+    /// Sets a new handler for this `InputBackend`
+    fn set_handler<H: InputHandler + 'static>(&mut self, handler: H);
+    /// Get a reference to the currently set handler, if any
+    fn get_handler(&mut self) -> Option<&mut InputHandler>;
+    /// Clears the currently handler, if one is set
+    fn clear_handler(&mut self);
+    /// Sets the cursor position, useful for e.g. pointer wrapping.
+    ///
+    /// Not guaranteed to be supported on every backend. The result usually
+    /// depends on the capability of the backend, but may also fail for certain
+    /// specific actions. See the backends documentation.
+    fn set_cursor_position(&mut self, x: u32, y: u32) -> Result<(), ()>;
+}
+
+/// Implement to receive input events from any `InputBackend`.
+pub trait InputHandler {
+    /// Called when a new `Seat` has been created
+    fn on_seat_created(&mut self, seat: &Seat);
+    /// Called when an existing `Seat` has been destroyed.
+    fn on_seat_destroyed(&mut self, seat: &Seat);
+    /// Called when a new keyboard event was received.
+    ///
+    /// # Arguments
+    ///
+    /// - `seat` - The `Seat` the event belongs to
+    /// - `time` - A upward counting variable useful for event ordering. Makes no gurantees about actual time passed between events.
+    /// - `key_code` - Code of the pressed key. See linux/input-event-codes.h
+    /// - `state` - `KeyState` of the event
+    /// - `count` - Amount of key presses registered
```

Comment:
Bad wording. That is the number of total keys pressed on the whole seat.
See https://wayland.freedesktop.org/libinput/doc/latest/group__event__keyboard.html#ga75a647f08ef9df9db7a6ebb9b297add4